lib/Dialect/TritonGPU/Transforms/Pipeline.cpp

```diff
@@ -8,23 +8,86 @@
 #include "triton/Analysis/Utility.h"
 #include "triton/Dialect/TritonGPU/IR/Dialect.h"
 #include "triton/Dialect/TritonGPU/Transforms/Passes.h"
+#include "llvm/ADT/MapVector.h"
 
 //===----------------------------------------------------------------------===//
+// This file implements software pipelining for loops. The implementation here
+// is inspired by the pipeline pass in Triton (version 2.0) and SCF's
+// LoopPipelining.
 //
-// This file implements loop software pipelining
-// The implementation here is inspired by the pipeline pass in Triton (-v2.0)
-// and SCF's LoopPipelining.
+// We divide the loop body into the following phases:
+// a. Pre-load operations: for instance, index computation.
+// b. Load operations: loading from global memory to shared memory.
+// c. Compute operations: for instance, Triton dot.
+// d. Post-load operations: for instance, index computation.
+//
+// To pipeline the loop, we need to:
+// - Hoist the load operations out of the loop body.
+// - Find all the dependencies of the load operations.
+// - Repeat the dependencies (numStage-1) times.
```

Comment:
would the following be better?
- Rematerialize the dependencies for their values at the first `numStage - 1` iterations